### DISCUSSION TITLE
First time in the Linux kernel development and kworkflow? Then, start your journey here :)

### DISCUSSION THREAD

**Discussion by rodrigosiqueira** [kworkflow/kworkflow#476] [Announcements]
Hi,

Are you interested in contributing to kworkflow and learn more about Linux kernel development? Are you a newcomer to these projects? If yes (or partially "yes") to any of these questions, this page is for you! My goal here is to help people to start working with kw in an enjoyable way. In the next set of sections, you will see a couple of steps that I recommend you to follow in order. 

# Phase 1: Buiding the fundamentals knowledge about Linux Kernel

If you want to contribute to kw, I highly recommend you to learn at least how to play with the Linux kernel by following some of the below steps.

## Step 1: Create a QEMU virtual machine

If you have a virtual machine, you can safely play with the Linux kernel code, and you can also use kw for that. For example, kw supports three types of target: vm, remote, and local; if you have a virtual machine, you can easily test these three features. In this sense, I highly recommend you to follow this tutorial:

&#8594; https://flusp.ime.usp.br/kernel/use-qemu-to-play-with-linux/

You can consider this step as done if you:
* Create two virtual machines, one with ArchLinux and another with Debian;
* You must be able to ssh in the machine;
* The shared folder between the QEMU VM and your host machine must work perfectly;
* If you understand every step of the tutorial.

Final recommendations for this part:

* Create an image with at least 13GB and uses qcaw2 format;
* Make a backup of your image because if you crash your working VM, you can easily copy your original VM and start work again without waste time setup everything again;

## Step 2: Compile and install Linux kernel (manually)

In this step, I want you to learn the basic set of commands required for compiling and installing the Linux kernel in your VM. The following tutorial is very detailed, and It uses the "hard way," which is the best way for learning:

&#8594; https://flusp.ime.usp.br/kernel/Kernel-compilation-and-installation/

You can consider this step done if you:
* You feel capable of compile the Linux kernel;
* Understand the importance of the config file;
* Know how to optimize your config file for a fast compilation;
* Installed your custom kernel in your VM;
* If you know how to remove an installed kernel in your VM;

## Step 3 (optional): Understand a little bit about the modules

This step is not mandatory but highly recommended. After the above steps, I recommend that you learn the basics about kernel modules by following this page:

&#8594; https://flusp.ime.usp.br/kernel/play_with_modules/

Acceptance criteria:

* Understand what a module is;
* Know how to install a module;
* Know how to load and unload a module;

Extra:

* If you want to learn a little bit more about one specific driver, take some time to read:
 &#8594; https://flusp.ime.usp.br/iio/experiment-one-iio-dummy/

# Phase 2: Start playing with kw

Ok, now you know how to compile and install the Linux kernel from its source. Now it is time to learn how to use kw to automate most of the manual steps you performed in the previous phase.

## Step 1: Installing kw

&#8594; https://kworkflow.org/tutorials/setup.html

## Step 2: Learn how to manage your config file

&#8594; https://kworkflow.org/tutorials/configm.html

## Step 3: Learn how to compile for an x86 architecture

&#8594; https://kworkflow.org/tutorials/buildlinux.html

## Step 4: How about use kw to manage your progress?

&#8594; https://kworkflow.org/tutorials/pomodoro-report.html

# Phase 3: Learn about Bash

Have you ever dedicated a few hours to learn about Bash? I mean, you probably spent some time studying your favorite programming language by reading books and doing tons of exercise. How about devoting some time to learn Bash? In my humble opinion, Bash documentation is a masterpiece since it is short and very well written; for this reason, I strongly recommend reading Bash manual:

&#8594; https://www.gnu.org/software/bash/manual/bash.pdf

From the above link, notice how short it is the Bash documentation. In particular, try your best to read from chapters 1 to 7, and pay special attention to chapter 3.

# Phase 4: Learn how to contribute to kw

Ok, you probably learned a lot at this point, and you want to contribute to kw. I recommend you to read the following pages:

1. How to contribute: https://kworkflow.org/content/howtocontribute.html
2. Development workflow: https://kworkflow.org/content/developmentworkflow.html
3. About tests: https://kworkflow.org/content/tests.html
4. About code style: https://kworkflow.org/content/codingstyle.html
5. Good first issue: https://github.com/kworkflow/kworkflow/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22

# Conclusion

If you are a student and completed all the above phases, you probably have good knowledge about Linux kernel development and Bash. Hopefully, you will become a kernel developer that uses kw in your daily tasks; hopefully, you will send some patches to us to make things better for everyone.

**Comment by paulormm** [comment]
Links from FLUSP website changed (and they are wrong here). The updated links are:

> 
> ## Step 1: Create a QEMU virtual machine
> 

https://flusp.ime.usp.br/kernel/use-qemu-to-play-with-linux/

> 
> ## Step 2: Compile and install Linux kernel (manually)
> 

https://flusp.ime.usp.br/kernel/Kernel-compilation-and-installation/

> 
> ## Step 3 (optional): Understand a little bit about the modules
> 

https://flusp.ime.usp.br/kernel/play_with_modules/

> 
> Extra:
> 

https://flusp.ime.usp.br/iio/experiment-one-iio-dummy/

**Reply by kwy95** [reply]
Links were updated, thank you.

**Comment by sahil-sagwekar2652** [comment]
The link for "Step 2: Learn how to manage your config file" is pointing to an older version which is incompatible with the latest.
The correct link should be - https://kworkflow.org/tutorials/kernel-config-manager.html

**Reply by paulormm** [reply]
@davidbtadokoro May you update it in this discussion/steps?

**Comment by ZXAfromOVERWORLD** [comment]
i would like to work on this project

**Reply by ZXAfromOVERWORLD** [reply]
currently completing phase 1 then i'll get in touch

**Comment by Cyber141k** [comment]
Done with phase 1 with arch linux
there are some errors also
